### ISSUE TITLE
[Bug]: No way to prevent Flatpak apps from accessing ~/.local/share/fonts

### ISSUE THREAD

**Issue by FaridZelli** [flatpak/flatpak#6045]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.12

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

41

### What architecture are you using?

x86_64

### Expected Behavior

Apps should not be able to read the contents of `~/.local/share/fonts` while `flatpak info --file-access` returns `hidden`.

### Actual Behavior

Able to read contents of `~/.local/share/fonts` in LibreOffice and Firefox even while `flatpak info --file-access` returns `hidden`.   

Example No. 1:
```
@localhost ~ $ flatpak override -u --reset org.mozilla.firefox

@localhost ~ $ flatpak info --file-access=/home/user/.local/share/fonts/Roboto-Regular.ttf org.mozilla.firefox
hidden
```
Example No. 2:
```
@localhost ~ $ flatpak override -u --reset org.libreoffice.LibreOffice

@localhost ~ $ flatpak info --file-access=/home/user/.local/share/fonts/Roboto-Regular.ttf org.libreoffice.LibreOffice
read-write

@localhost ~ $ flatpak override -u --nofilesystem=/home/user/.local/share/fonts/ org.libreoffice.LibreOffice

@localhost ~ $ flatpak info --file-access=/home/user/.local/share/fonts/Roboto-Regular.ttf org.libreoffice.LibreOffice
hidden
```

Outcome:   
- `Roboto-Regular.ttf` remains accessible in both examples. The only way to prevent the font from being detected is to delete it from `~/.local/share/fonts` entirely.

### Additional Information

Potentially related to #1781 (however, not a duplicate)

Q: Are there any real-world implications for this?
A: Hiding custom fonts significantly reduces the effectiveness of [font fingerprinting](https://undetectable.io/blog/fonts-browser-fingerprints/) in various web applications and browsers.

**Comment by chrisawi** [comment]
Flatpak provides unconditional access to several directories where doing so doesn't compromise safety or robustness, e.g. user & system fonts and user & system icons. These are all mounted inside `/run/host/` inside the sandbox. Filesystem overrides don't affect them because they only apply to directories that are 'exported' at their actual location inside the sandbox. That's also why `--file-access` doesn't know about them.

> A: Hiding custom fonts significantly reduces the effectiveness of [font fingerprinting](https://undetectable.io/blog/fonts-browser-fingerprints/) in various web applications and browsers.

IMHO, that's not really in scope for Flatpak. Flatpak apps can trivially fingerprint users (`/etc/machine-id` for one) if they want to. It's up to those apps to try to prevent fingerprinting by third parties. Trying to constrain them with the Flatpak sandbox for this purpose seems misguided. See also https://github.com/flatpak/flatpak/issues/4311.

**Comment by FaridZelli** [comment]
> Flatpak provides unconditional access to several directories where doing so doesn't compromise safety or robustness, e.g. user & system fonts and user & system icons. These are all mounted inside /run/host/ inside the sandbox.

> IMHO, that's not really in scope for Flatpak.

The user should have the ability to override access to these directories regardless. If Flatpak is hardcoding these rules, it must offer a way to override them.

Additionally, I was unable to find any documentation on these hardcoded directories. It should be clearly noted that the sandbox does **not** fully limit access to the user's home directory when instructed to do so.

From flatpak-override(1) --nofilesystem:

> This option does not prevent access to a more narrowly-scoped --filesystem. For example, if an application has the equivalent of --filesystem=xdg-config/MyApp in its manifest or as a system-wide override, and flatpak override --user --nofilesystem=home as a per-user override, then it will be prevented from accessing most of the home directory, but it will still be allowed to access $XDG_CONFIG_HOME/MyApp.

It is nowhere stated that Flatpak has its own predefined system-wide overrides, nor what they are.

Additionally, `flatpak info --file-access` should serve as a reliable way to check filesystem access permissions for apps, and must under no circumstances return `hidden` when an app does have access to the contents of the directory, in one way or another.

**Comment by chrisawi** [comment]
> The user should have the ability to override access to these directories regardless. If Flatpak is hardcoding these rules, it must offer a way to override them.

I don't think we can presuppose that. Unless there is a specific and relevant problem introduced by this feature and the other automatic members of `/run/host`, there doesn't inherently need to be a mechanism to disable it.

> I was unable to find any documentation on these hardcoded directories. It should be clearly stated that the sandbox does not fully limit access to the user's home directory when instructed to do so.

You're right that this should be documented. A natural spot might be in `flatpak-run(1)` next to where it mentions `/run/host/os-release`.

> It is nowhere mentioned that Flatpak has its own predefined system-wide overrides, nor what they are.

In the interest of clear terminology, it's not an 'override'. Overrides are user-made changes to the static permissions set in an app's metadata. If Flatpak were to add an opt-out (e.g. `--no-host-integration`), then *that* could be set as an override.

> Additionally, `flatpak info --file-access` should serve as a reliable way to check filesystem access permissions for apps, and must under no circumstances return `hidden` when an app does have access to the contents of the directory, in one way or another.

`--file-access` indicates that a file is available inside the sandbox at that specific path. This is API that can't be changed. For one, it's used by the document portal to avoid proxying sandbox-accessible files. If `--file-access` were to return `read-only` for a path in `~/.local/share/fonts`, that could cause the app to receive a host path that it can't use.

**Comment by FaridZelli** [comment]
> If Flatpak were to add an opt-out (e.g. `--no-host-integration`), then _that_ could be set as an override.

This would be perfect. Thank you for taking the time to clarify things. Adding `--no-host-integration` sounds like the way to go.

**Comment by FaridZelli** [comment]
> Fine-grained control over the various things that integrate with mechanisms on the host (for example a finer-grained option like --no-host-fonts) sounds like a lot of code (therefore a lot of bugs)

You make a good point here, and I agree that we need an elegant solution to prevent complications in the future. It would be unnecessary to include flags for every directory, as the user can simply add overrides for them manually.

> In particular, if you're running a GUI app, then you're implicitly expecting it to have sufficient host integration to be able to contact the Wayland or X11 socket on the host - without that, it simply wouldn't work.

> turning off any of the host-integration mechanisms is essentially a "make my app stop working" switch.

This is also a valid point. However, the concern here is that **Flatpak, by default, whitelists a sparsely documented list of directories** for this purpose. As an advanced user, I should be informed of this behaviour, and have the option to further isolate an app if I desire to do so. What I'm asking here is not for Flatpak to support off-label use cases, but to offer more flexibility and control over the sandbox.
